File: golang/11.binary-tree/main/count_nodes.go
package main

import (
	"github.com/AmirAhmadzadeh/problems/binary_tree"
	"github.com/AmirAhmadzadeh/problems/node"
)

// bfs
func countNodes1(bt *binary_tree.BinaryTree) int {
	root := bt.GetRoot()
	queue := []*node.Node{root}
	seen := make(map[int]bool)
	result := 0

	for len(queue) > 0 {
		vertex := queue[0]
		queue = queue[1:]

		if !seen[vertex.Value()] {
			seen[vertex.Value()] = true
			result += 1
		}

		if vertex.Right() != nil {
			queue = append(queue, vertex.Right())
		}

		if vertex.Left() != nil {
			queue = append(queue, vertex.Left())
		}
	}

	return result
}

/// dfs
func dfsCount(result int, seen map[int]bool, queue *[]*node.Node) int {
	if len(*queue) == 0 {
		return result
	}

	vertex := (*queue)[0]
	*queue = (*queue)[1:]

	seen[vertex.Value()] = true
	result += 1

	if vertex.Right() != nil {
		*queue = append(*queue, vertex.Right())
	}

	if vertex.Left() != nil {
		*queue = append(*queue, vertex.Left())
	}

	return dfsCount(result, seen, queue)
}

func countNodes2(bt *binary_tree.BinaryTree) int {
	root := bt.GetRoot()
	seen := make(map[int]bool)
	queue := []*node.Node{root}
	result := 0

	return dfsCount(result, seen, &queue)
}

File: ts/11.binary-tree/countNodes.ts
import { Node } from './';

export function isNodeExists<T>(
    node: Node<T> | null,
    indexToFind: number,
    height: number
) {
    let left = 0;
    let right = 2 ** height - 1;

    for (let i = 0; i < height; i++) {
        let mid = Math.ceil((left + right) / 2);
        if (indexToFind >= mid) {
            node = node?.right as Node<T> | null;
            left = mid;
        } else {
            node = node?.left as Node<T> | null;
            right = mid - 1;
        }
    }

    return node !== null;
}

export function getCompleteTreeHeight<T>(node: Node<T> | null): number {
    if (!node) return 0;
    let height = 0;

    while (node.left) {
        height += 1;
        node = node.left;
    }
    return height;
}

export function countNodes<T>(root: Node<T> | null) {
    if (!root) return 0;

    let height = getCompleteTreeHeight({ ...root });
    if (height === 0) return 1;

    let result = 2 ** height - 1;

    let left = 0;
    let right = 2 ** height - 1;

    while (left < right) {
        let mid = Math.ceil((left + right) / 2);
        if (isNodeExists({ ...root }, mid, height)) {
            left = mid;
        } else {
            right = mid - 1;
        }
    }

    return result + left + 1;
}
